grails-app/services/org/pih/warehouse/invoice/PrepaymentInvoiceService.groovy

```diff
@@ -23,47 +57,222 @@ class PrepaymentInvoiceService {
+        // Multiplied by (-1) to keep inverse items as negative amount
+        inverseItem.amount = quantity * prepaymentItem.unitPrice * prepaymentPercent * (-1)
+        return inverseItem
+    }
 
-        order.activeOrderItems.each { OrderItem orderItem ->
-            InvoiceItem invoiceItem = invoiceService.createFromOrderItem(orderItem)
-            invoice.addToInvoiceItems(invoiceItem)
+    private InvoiceItem createInverseItemForOrderAdjustment(OrderAdjustment orderAdjustment) {
+        InvoiceItem prepaymentItem = orderAdjustment.invoiceItems.find { it.isPrepaymentInvoice }
+        InvoiceItem inverseItem = createFromOrderAdjustment(orderAdjustment)
+        // For order adjustment invoiceItem.quantity is 1 or 0, but for inverse should be for now 1
+        inverseItem.quantity = 1
+        inverseItem.invoiceItemType = InvoiceItemType.INVERSE
+        inverseItem.unitPrice = prepaymentItem.unitPrice
+        // Multiplied by (-1) to keep inverse items as negative amount
+        inverseItem.amount = prepaymentItem.amount * (-1)
+        return inverseItem
+    }
+
+    private Integer getQuantityAvailableToInverse(OrderItem orderItem, InvoiceItem prepaymentItem) {
+        Integer quantityInversed = orderItem.allInvoiceItems.findAll { it.inverseItem }.sum { it.quantity } ?: 0
+        return prepaymentItem.quantity > quantityInversed ? prepaymentItem.quantity - quantityInversed : 0
+    }
+
+    /**
+     * Updates invoice item's quantity and amount and finds related inverse item and updates quantity and amount
+     * */
+    void updateInvoiceItemQuantity(String itemId, Integer quantity) {
+        InvoiceItem invoiceItem = InvoiceItem.get(itemId)
+        if (quantity > invoiceItem.quantity) {
+            throw new IllegalArgumentException("Cannot update quantity invoiced to higher value")
         }
+        if (quantity <= 0) {
+            throw new IllegalArgumentException("Quantity invoiced needs to be higher than 0")
+        }
+        // I think this entire update is only for shipment items (but will double check that)
+        if (!invoiceItem?.shipmentItem) {
+            throw new IllegalArgumentException("Invoice item is missing shipment item. Cannot edit quantity.")
+        }
+        // update invoice item's quantity (and adjust amount)
+        invoiceItem.quantity = quantity
+        invoiceItem.amount = quantity * invoiceItem.unitPrice
+        // update inverse item's quantity (and adjust amount)
+        OrderItem orderItem = invoiceItem?.shipmentItem?.orderItems?.find { it }
+        BigDecimal prepaymentPercent = (orderItem.order.paymentTerm?.prepaymentPercent ?: Constants.DEFAULT_PAYMENT_PERCENT) / 100
+        InvoiceItem prepaymentItem = orderItem.invoiceItems.find { it.isPrepaymentInvoice }
+        InvoiceItem inverseItem = findInverseItem(invoiceItem)
+        inverseItem.quantity = quantity
+        inverseItem.amount = quantity * prepaymentItem.unitPrice * prepaymentPercent * (-1)
```

Comment:
can you move this amount calculation to a method so this logic doesn't need to be duplicated everywhere?

Reply:
Sure will do it. btw I am moving out the edit endpoint to separate PR.